TrainingExtensions/torch/src/python/aimet_torch/quantsim_straight_through_grad.py

```diff
@@ -98,6 +101,80 @@ def _broadcast_to_tensor(tensor, encoding):
     return dloss_by_dx
 
 
+def compute_dloss_by_dmin_using_dmax(dloss_by_dmax):
+    """
+    compute derivative of loss w.r.t min, it is sign flipped version of derivative w.r.t max
+    :param dloss_by_dmax derivative w.r.t max
+    :return: derivative of loss w.r.t min
+    """
+
+    return -1 * dloss_by_dmax
+
+def compute_dloss_by_dmax(x, grad, scaling, offset, n, p, ch_axis=0):
+    """
+    helper function to compute derivative of loss w.r.t encoding max
+    :param x: input
+    :param grad: gradient
+    :param scaling: scaling factor computed for given encoding min/max
+    :param offset: offset computed
+    :param n: lower bound
+    :param p: upper bound
+    :param ch_axis: channel axis along which sum is computed for gradient calculation
+    :return: computed derivative of loss w.r.t encoding max
+    """
+
+    r_x_by_s_plus_round_o = torch.round(x / scaling) + torch.round(offset)
+    # R(x/s)-(x/s)
+    r_x_by_s_minus_x_by_s = torch.round(x / scaling) - (x / scaling)
+
+    # compute dq_by_dmax
+    # expr to be used if r_x_by_s_plus_round_o < n or > p
+    false_expr = torch.clamp(r_x_by_s_plus_round_o.data, min=n.data.item(), max=p.data.item()) * (1 / p)
+    inner_cond = torch.where(torch.le(r_x_by_s_plus_round_o.data, p), (r_x_by_s_minus_x_by_s * (1 / p)), false_expr)
+
+    # we need a scalar value for dq_by_dmax, so reduce 4d value computed above
+    # to single value before returning gradient
+    # this uses chain rule, multiply by loss and sum it to get scalar.
+    dq_by_dmax = torch.where(torch.le(n, r_x_by_s_plus_round_o.data), inner_cond, false_expr)
+
+    dloss_by_dmax = dq_by_dmax * grad
+    if len(scaling) > 1 and len(x.shape) > 1:
```

Comment:
What happens if len(scaling) > 1 but len(x.shape) is not > 1? If this is invalid case, we should assert. It feels like this if/elif is missing some corner case handling.